### PR TITLE
Support Mellanox NIC (mlx) tuning

### PR DESCRIPTION
**Pull request by tarzanek**
Mellanox NICs are used in Azure cloud, we need to add comparator that can detect their IRQs and query IDs in interrupts so we can properly tune them.
Also Azure has a notion of exposing the physical nic for tuning, so perftune has to be able to tune both physical and its virtual interface in one run, so we shall add support for multiple nics to be tuned.

fixes https://github.com/scylladb/scylla/issues/6266

### CODE REVIEW

**Comment by tarzanek** [comment]
@vladzcloudius can you please review?

**Comment by tarzanek** [comment]
hmm I am actually not sure now if 2 is a good number, but for sq split mode (which selects 2 cpus) this is a good default, so just need to make sure that this happens only in this case
(that said I only did positive testing on simple L32s, so I need more tests on mlx5 and definitely a negative case)

**Comment by tarzanek** [comment]
btw. I noticed that Azure keeps phys iface names as enPxxxx (e.g. enP34867s1)
however after stop (deallocate) and run the name will change (and nvmes can flush of course), so
we shall be careful on Azure when the name changes ... but I guess making perftune error out(and scylla stop starting) will be enough of notice

**Comment by dorlaor** [comment]
Cheers @tarzanek

File: scripts/perftune.py [hunk 1]
```diff
@@ -487,11 +487,11 @@ def tune(self):
             self.__setup_bonding_iface()
 
         # Increase the socket listen() backlog
-        fwriteln_and_log('/proc/sys/net/core/somaxconn', '4096')
+        fwriteln_and_log('/proc/sys/net/core/somaxconn', '16384')
```
Comment (vladzcloudius) [inline scripts/perftune.py:490]: This change is not related to MLX cards - please, put it a separate patch with a corresponding description.

File: scripts/perftune.py [hunk 2]
```diff
@@ -487,11 +487,11 @@ def tune(self):
...
 
         # Increase the maximum number of remembered connection requests, which are still
         # did not receive an acknowledgment from connecting client.
-        fwriteln_and_log('/proc/sys/net/ipv4/tcp_max_syn_backlog', '4096')
+        fwriteln_and_log('/proc/sys/net/ipv4/tcp_max_syn_backlog', '16384')
```
Comment (vladzcloudius) [inline scripts/perftune.py:494]: same here

File: scripts/perftune.py [hunk 3]
```diff
@@ -672,7 +691,12 @@ def __learn_irqs_one(self, iface):
         fp_irqs_re = re.compile("\-TxRx\-|\-fp\-|\-Tx\-Rx\-|mlx\d+\-\d+@")
         irqs = list(filter(lambda irq : fp_irqs_re.search(self.__irqs2procline[irq]), all_irqs))
         if irqs:
-            irqs.sort(key=self.__intel_irq_to_queue_idx)
+            driver_name = self.__get_driver_name(iface)
+            print("Sorting IRQs by driver: {}".format(driver_name))
```
Comment (vladzcloudius) [inline scripts/perftune.py:695]: Use `perftune_print(...)`

File: scripts/perftune.py [hunk 4]
```diff
@@ -672,7 +691,12 @@ def __learn_irqs_one(self, iface):
...
+            if (driver_name.startswith("mlx")):
+                irqs.sort()
```
Comment (vladzcloudius) [inline scripts/perftune.py:697]: I'm not sure that you can assume that IRQs numbers ordering is the same as corresponding HW queues order. 
I was sorting HW queues by their index for Intel NICs for exactly same reason.

File: scripts/perftune.py [hunk 5]
```diff
@@ -705,6 +729,7 @@ def __setup_one_hw_iface(self, iface):
         #
         # If this NIC has a limited number of Rx queues then we want to distribute their IRQs separately.
         # For such NICs we've sorted IRQs list so that IRQs that handle Rx are all at the head of the list.
+        print("Number of Rx queues: {}".format(max_num_rx_queues))
```
Comment (vladzcloudius) [inline scripts/perftune.py:732]: `perftune_print`

File: scripts/perftune.py [hunk 6]
```diff
@@ -738,21 +763,12 @@ def __max_rx_queue_count(self, iface):
         ixgbevf: VF NICs support up to 4 RSS queues.
         i40e:    PF NICs support up to 64 RSS queues.
         i40evf:  VF NICs support up to 16 RSS queues.
+        mlx:     PF/VF NICs throttle to 2 due to si problems when cpu is shared with scylla
```
Comment (vladzcloudius) [inline scripts/perftune.py:766]: You want to reduce the number of HW queues to the number of IRQ CPUs.
It's not always 2.

File: scripts/perftune.py [hunk 7]
```diff
@@ -738,21 +763,12 @@ def __max_rx_queue_count(self, iface):
...
-        ethtool_i_lines = run_read_only_command(['ethtool', '-i', iface]).splitlines()
-        driver_re = re.compile("driver:")
-        driver_lines = list(filter(lambda one_line: driver_re.search(one_line), ethtool_i_lines))
-
-        if driver_lines:
-            if len(driver_lines) > 1:
-                raise Exception("More than one 'driver:' entries in the 'ethtool -i {}' output. Unable to continue.".format(iface))
-
-            driver_name = driver_lines[0].split()[1].strip()
+        driver_to_max_rss = {'ixgbe': 16, 'ixgbevf': 4, 'i40e': 64, 'i40evf': 16, 'mlx4_en': 2,'mlx5_core' : 2}
```
Comment (vladzcloudius) [inline scripts/perftune.py:769]: You seem to forget to actually call `ethtool -L` to reduce the number of RX queues.
Please, don't forget to test your code before sending a PR.

Comment (tarzanek) [reply scripts/perftune.py:769]: yes, correct, this is was missed, ethtool -L
(and I tested IRQs but forgot to check number of queues)
will fix

File: scripts/perftune.py (hunk 6 above)
Comment (tarzanek) [reply scripts/perftune.py:766]: correct, will be dynamic per cpu tuning

File: scripts/perftune.py (hunk 4 above)
Comment (tarzanek) [reply scripts/perftune.py:697]: what is the source of truth for ordering?
parse queue ids and assign them to positions by their numbers?
if yes, then this is OK, I will do it like this, but if this is the same for intel, then I guess I will change parsing code to work slightly better

Comment (vladzcloudius) [reply scripts/perftune.py:697]: The index in the name of the queue in `/proc/interrupts`

File: scripts/perftune.py (hunk 6 above)
Comment (vladzcloudius) [reply scripts/perftune.py:766]: Note that this may be any number since we have `--irq-cpu-mask`

**Comment by tarzanek** [comment]
add:
add a boolean param to enable the logic of changing # of cpus for irq
this switch will leverage --irq-cpu-mask
check NIC # of queues first, distribute them per mask

the logic will simply change RX queues and distribute them appropriately if enabled

**Comment by tarzanek** [comment]
note: if there is no filter for irqs for queue for given card, code should fail

**Comment by tarzanek** [comment]
add also support for tuning of virtual device when on top of physical (in case of advanced networking) - in a similar way like we do bond tuning (so tune sw tunables on virtual interface)

**Comment by vladzcloudius** [comment]
> add also support for tuning of virtual device when on top of physical (in case of advanced networking) - in a similar way like we do bond tuning (so tune sw tunables on virtual interface)

We don't need the notion of "SW on top of HW NIC" inside the perftune.py itself.
We need a way to give it multiple NIC interfaces and make perftune.py configure all of them.
In that case if the configuration mode is not provided the least of the modes supported by provided NICs should be chosen and configured on all those NICs (and other HW configured in the same script invocation).

The already apply the same logic when we tune HW NICs comprising the bond interface when the script is request to tune that bond interface.

Also there is not need to special-case the logical interface. More than that - no need to even try to detect which iface is logical and which one is physical.
Simply configure all given interfaces in an absolutely standard way.

File: scripts/perftune.py [hunk 8]
```diff
@@ -643,6 +643,35 @@ def __intel_irq_to_queue_idx(self, irq):
         else:
             return sys.maxsize
 
+    def __mlx_irq_to_queue_idx(self, irq):
+        """
+        Return the HW queue index for a given IRQ for Intel NICs in order to sort the IRQs' list by this index.
```
Comment (vladzcloudius) [inline scripts/perftune.py:648]: The comment refers Intel which the function name implies Mellanox.

Comment (vladzcloudius) [reply scripts/perftune.py:648]: It looks half baked anyway - I should probably wait with a review till it's ready... ;)

File: scripts/perftune.py (hunk 7 above)
Comment (tarzanek) [reply scripts/perftune.py]: I used 8 and 32 as max for mellanox 4 and 5 drivers, this however should be validated with mellanox docs, I lack access, so I need help from @vladzcloudius or @eyalgutkind to help me validate these numbers (observed from VMs or from public docs, which are not definitive)

File: scripts/perftune.py [hunk 9]
```diff
@@ -643,6 +643,77 @@ def __intel_irq_to_queue_idx(self, irq):
...
             return sys.maxsize
 
+    def __mlx_irq_to_queue_idx(self, irq):
+        """
+        Return the HW queue index for a given IRQ for Mellanox NICs in order to sort the IRQs' list by this index.
+
+        Mellanox NICs have the IRQ which name looks like
+        this:
+        mlx5_comp23
+             <bla-bla>mlxX_comp<index>
```
Comment (vladzcloudius) [inline scripts/perftune.py:653]: What <bla-bla> stands for here?

File: scripts/perftune.py [hunk 10]
```diff
@@ -643,6 +643,77 @@ def __intel_irq_to_queue_idx(self, irq):
...
+        or this:
+        mlx4-6
+             <bla-bla>mlxX-<index>
+
+        :param irq: IRQ number
+        :return: HW queue index for Mellanox NICs and 0 for all other NICs
+        """
+        mlx_fp_irq_re = re.compile("mlx\d_comp(\d+)|mlx\d\-(\d+)")
```
Comment (vladzcloudius) [inline scripts/perftune.py:661]: Why `mlx\d_comp` ? It should be `mlx5_comp`, no?
The same about `mlx4` pattern.

File: scripts/perftune.py [hunk 11]
```diff
@@ -643,6 +643,77 @@ def __intel_irq_to_queue_idx(self, irq):
...
+
+        m = mlx_fp_irq_re.search(self.__irqs2procline[irq])
+        queue=m.group(2)
+        return int(queue)
+
+    def __get_driver_name(self, iface):
```
Comment (vladzcloudius) [inline scripts/perftune.py:667]: What is this? How is this related to Mlx support?
Move to a separate patch, please.

File: scripts/perftune.py [hunk 12]
```diff
@@ -643,6 +643,77 @@ def __intel_irq_to_queue_idx(self, irq):
...
+        return driver_name
+
+    def __set_driver_rx_queues(self, iface, queue_count):
+        """
+        :param iface: interface to set
+        :param queue_count: number of RX queues to set
+        """
+        try:
+                perftune_print("Limiting RX queues on {} ...".format(iface))
+                run_one_command(['ethtool','-L', iface, 'rx', str(queue_count)], stderr=subprocess.DEVNULL)
```
Comment (vladzcloudius) [inline scripts/perftune.py:693]: Sometimes you need to set `combined` in order to set `rx` queues count. - Many vendors don't allow configuring different Rx and Tx queues number. Therefore they use `combined` value.

Look in the `man ethtool`

File: scripts/perftune.py [hunk 13]
```diff
@@ -705,12 +784,22 @@ def __setup_one_hw_iface(self, iface):
...
             perftune_print("Distributing IRQs handling Rx:")
             distribute_irqs(all_irqs[0:num_rx_queues], self.irqs_cpu_mask)
             perftune_print("Distributing the rest of IRQs")
             distribute_irqs(all_irqs[num_rx_queues:], self.irqs_cpu_mask)
+
+            #use ethtool to tune down rx queues if this is mlx interface
+            driver_name = self.__get_driver_name(iface)
+            if (driver_name.startswith("mlx")):
+                queues_count_from_cpuset = bin(int(self.irqs_cpu_mask,16)).count("1")
+                qmax,qcurrent=self.__get_current_rx_queues(iface)
```
Comment (vladzcloudius) [inline scripts/perftune.py:799]: Space after a coma, please.

File: scripts/perftune.py [hunk 14]
```diff
@@ -705,12 +784,22 @@ def __setup_one_hw_iface(self, iface):
...
             num_rx_queues = self.__get_rx_queue_count(iface)
             perftune_print("Distributing IRQs handling Rx:")
             distribute_irqs(all_irqs[0:num_rx_queues], self.irqs_cpu_mask)
             perftune_print("Distributing the rest of IRQs")
             distribute_irqs(all_irqs[num_rx_queues:], self.irqs_cpu_mask)
+
+            #use ethtool to tune down rx queues if this is mlx interface
+            driver_name = self.__get_driver_name(iface)
+            if (driver_name.startswith("mlx")):
+                queues_count_from_cpuset = bin(int(self.irqs_cpu_mask,16)).count("1")
```
Comment (vladzcloudius) [inline scripts/perftune.py:798]: Space after a coma, please

File: scripts/perftune.py [hunk 15]
```diff
@@ -705,12 +784,22 @@ def __setup_one_hw_iface(self, iface):
...
+                qmax,qcurrent=self.__get_current_rx_queues(iface)
+                min_queue=queues_count_from_cpuset if qmax>queues_count_from_cpuset else qmax
```
Comment (vladzcloudius) [inline scripts/perftune.py:800]: `min_queue` is a very confusing name. `new_rx_queues_count` is better IMO.

File: scripts/perftune.py [hunk 16]
```diff
@@ -738,21 +827,12 @@ def __max_rx_queue_count(self, iface):
         ixgbevf: VF NICs support up to 4 RSS queues.
         i40e:    PF NICs support up to 64 RSS queues.
         i40evf:  VF NICs support up to 16 RSS queues.
+        mlx:     PF/VF NICs throttle to respective cpu count dedicated to sq split due to si problems when cpu is shared with scylla, this is minimum
```
Comment (vladzcloudius) [inline scripts/perftune.py:830]: Drop all changes in the `__max_rx_queue_count()` please. We don't know of any limitation in Mellanox NICs.

File: scripts/perftune.py [hunk 17]
```diff
@@ -832,7 +912,7 @@ def tune(self):
 
 #### Protected methods ##########################
     def _get_def_mode(self):
-        """ 
+        """
```
Comment (vladzcloudius) [inline scripts/perftune.py:915]: Don't mix cosmetic changes changes with the function change, please.

File: scripts/perftune.py [hunk 18]
```diff
@@ -1211,12 +1291,12 @@ def names():
...
 Default values:
 
- --nic NIC       - default: eth0
+ --nic NIC       - default: eth0 (you can specify multiple nics)
  --cpu-mask MASK - default: all available cores mask
  --tune-clock    - default: false
 ''')
 argp.add_argument('--mode', choices=PerfTunerBase.SupportedModes.names(), help='configuration mode')
-argp.add_argument('--nic', help='network interface name, by default uses \'eth0\'')
+argp.add_argument('--nic', action='append', help='network interface name(s), by default uses \'eth0\'')
```
Comment (vladzcloudius) [inline scripts/perftune.py:1299]: Allowing tuning multiple NICs should be a separate patch

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py (hunk 12 above)
Comment (tarzanek) [reply scripts/perftune.py:693]: ok, so that's why this is called __set_driver_rx_queues
later in the code it's only used with mlx driver prefix only (which support setting only rx, which is what we want)

do I need to add a check here for ethtool if it has combined or tx rx ?
(I think -l will just show channels, so I can detect if it supports rx and use rx, or do nothing)

File: scripts/perftune.py (hunk 16 above)
Comment (tarzanek) [reply scripts/perftune.py:830]: could this be a limit?
https://github.com/torvalds/linux/blob/master/drivers/net/ethernet/mellanox/mlx4/mlx4_en.h#L74

not sure where to look in https://github.com/torvalds/linux/tree/master/drivers/net/ethernet/mellanox/mlx5/core

anyways, dropping, we can always add a ceiling, the important is to lower it now to the dedicated cpus

File: scripts/perftune.py (hunk 18 above)
Comment (tarzanek) [reply scripts/perftune.py:1299]: it is, am I missing something? you mean like separate PR ?

File: scripts/perftune.py (hunk 1 above)
Comment (vladzcloudius) [inline scripts/perftune.py:490]: Please, put a motivation for this change in the patch description

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py (hunk 18 above)
Comment (vladzcloudius) [reply scripts/perftune.py:1299]: No. Probably I missed it.
Please, rephrase the patch description: the language is confusing. It reads as if you add a support for passing parameters for multiple NICs.

Something like "add a support for tuning multiple NICs in a single command" we be clearer.

File: scripts/perftune.py [hunk 19]
```diff
@@ -1211,12 +1211,12 @@ def names():
 
 Default values:
 
- --nic NIC       - default: eth0
+ --nic NIC       - default: eth0 (you can specify multiple nics)
  --cpu-mask MASK - default: all available cores mask
  --tune-clock    - default: false
 ''')
 argp.add_argument('--mode', choices=PerfTunerBase.SupportedModes.names(), help='configuration mode')
-argp.add_argument('--nic', help='network interface name, by default uses \'eth0\'')
```
Comment (vladzcloudius) [inline scripts/perftune.py:1219]: Please, change the args.nic to args.nics - it will make it more readable. Keep the `--nic` however: there is a way to define the destination field in the `args`. See how I did it for `--dir`

File: scripts/perftune.py [hunk 20]
```diff
@@ -1335,32 +1335,36 @@ def dump_config(prog_args):
...
-    if TuneModes.system.name in args.tune:
-        tuners.append(SystemPerfTuner(args))
-
-    # Set the minimum mode among all tuners
-    if not args.irq_cpu_mask:
-        mode = PerfTunerBase.SupportedModes.combine([tuner.mode for tuner in tuners])
-        for tuner in tuners:
-            tuner.mode = mode
+    nics=args.nic
+    for nic in nics:
```
Comment (vladzcloudius) [inline scripts/perftune.py:1419]: I think this is just wrong. 
What you do here is you simulate running perftune.py for each of given nics (and all the rest of parameters) one after another. 
While what you should be doing is to do the same as if given NICs are a part of a bonding interface.
Namely you should have calculated the combined tuning mode for ALL nics and the rest of tunables, and then tune everything in a single take.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py (hunk 12 above)
Comment (vladzcloudius) [reply scripts/perftune.py:693]: The driver can support either `rx + tx` or `combined`. In the later case setting "rx queues number" means setting the `combined` value. 

You can choose to implement setting `rx`, `tx` and `combined` in separate functions and then call them depending on the driver "mode" which you can detect.

Regardless, the current implementation (covering only non-combined case) is partial is not enough.

BTW, I do recall that one of mlx4 or mlx5 supports the former mode and another - the later.
So, even in the context of Mellanox cards you need to support both.

File: scripts/perftune.py (hunk 14 above)
Comment (vladzcloudius) [reply scripts/perftune.py:798]: I don't think you pushed your changes.

File: scripts/perftune.py (hunk 16 above)
Comment (vladzcloudius) [reply scripts/perftune.py:830]: Let me know when I can look at those changes.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py (hunk 9 above)
Comment (tarzanek) [reply scripts/perftune.py:653]: removed

File: scripts/perftune.py [hunk 21]
```diff
@@ -575,11 +581,11 @@ def __setup_rfs(self, iface):
         ethtool_msg = "Enable ntuple filtering HW offload for {}...".format(iface)
         if dry_run_mode:
                 perftune_print(ethtool_msg)
-                run_one_command(['ethtool','-K', iface, 'ntuple', 'on'], stderr=subprocess.DEVNULL)
+                run_ethtool(['-K', iface, 'ntuple', 'on'])
```
Comment (vladzcloudius) [inline scripts/perftune.py:588]: You didn't shut up the `stderr` as it was before.

Comment (vladzcloudius) [reply scripts/perftune.py:588]: In addition to that this patch changes the behaviour of the `dry run mode` - run_read_only_command() always executes a command while run_command only prints the command in a dry run mode.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py [hunk 22]
```diff
@@ -748,17 +767,7 @@ def __max_rx_queue_count(self, iface):
         """
         driver_to_max_rss = {'ixgbe': 16, 'ixgbevf': 4, 'i40e': 64, 'i40evf': 16}
 
-        driver_name = ''
-        ethtool_i_lines = run_read_only_command(['ethtool', '-i', iface]).splitlines()
```
Comment (vladzcloudius) [inline scripts/perftune.py:746]: This line change should have been covered in the previous patch.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]
This patch should have included only moving code from one place into another. But it's not (see my inline comments).

File: scripts/perftune.py [hunk 23]
```diff
@@ -649,6 +649,27 @@ def __intel_irq_to_queue_idx(self, irq):
...
+        or this:
+        mlx4-6
+             mlx4-<index>
+
+        :param irq: IRQ number
+        :return: HW queue index for Mellanox NICs and 0 for all other NICs
+        """
+        mlx_fp_irq_re = re.compile("mlx5_comp(\d+)|mlx4\-(\d+)")
+
+        m = mlx_fp_irq_re.search(self.__irqs2procline[irq])
```
Comment (vladzcloudius) [inline scripts/perftune.py:677]: Not good. What if search has failed?
See what I do in `__intel_irq_to_queue_idx` for that case.

File: scripts/perftune.py [hunk 24]
```diff
@@ -694,10 +718,15 @@ def __learn_irqs_one(self, iface):
         """
         # filter 'all_irqs' to only reference valid keys from 'irqs2procline' and avoid an IndexError on the 'irqs' search below
         all_irqs = set(learn_all_irqs_one("/sys/class/net/{}/device".format(iface), self.__irqs2procline, iface)).intersection(self.__irqs2procline.keys())
-        fp_irqs_re = re.compile("\-TxRx\-|\-fp\-|\-Tx\-Rx\-|mlx\d+\-\d+@")
+        fp_irqs_re = re.compile("\-TxRx\-|\-fp\-|\-Tx\-Rx\-|mlx4-\d+@|mlx5_comp\d+@")
         irqs = list(filter(lambda irq : fp_irqs_re.search(self.__irqs2procline[irq]), all_irqs))
         if irqs:
-            irqs.sort(key=self.__intel_irq_to_queue_idx)
+            driver_name = self.__get_driver_name(iface)
+            perftune_print("Sorting IRQs by driver: {}".format(driver_name))
```
Comment (vladzcloudius) [inline scripts/perftune.py:776]: This printout may be confusing when the NIC vendor is neither Mellanox nor Intel (e.g. Broadcom).
In that case you'll not be sorting by the `bnx2x` driver like the printout would suggest.

Let's change the printout to be "Driver X has been detected" and that's it. No need to say anything about the sorting.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py [hunk 25]
```diff
@@ -759,12 +802,24 @@ def __setup_one_hw_iface(self, iface):
...
             num_rx_queues = self.__get_rx_queue_count(iface)
             perftune_print("Distributing IRQs handling Rx:")
             distribute_irqs(all_irqs[0:num_rx_queues], self.irqs_cpu_mask)
             perftune_print("Distributing the rest of IRQs")
             distribute_irqs(all_irqs[num_rx_queues:], self.irqs_cpu_mask)
+
+            #use ethtool to tune down rx queues if user wants to do it
+            if self.args.tune_rxqueues:
+                queues_count_from_cpuset = bin(
+                    int(self.irqs_cpu_mask, 16)).count("1")
```
Comment (vladzcloudius) [inline scripts/perftune.py:828]: Let's make it a single line, please.

File: scripts/perftune.py [hunk 26]
```diff
@@ -759,12 +802,24 @@ def __setup_one_hw_iface(self, iface):
...
+                qmax, qcurrent = self.__get_current_rx_queues(iface)
+                new_rx_queues_count = queues_count_from_cpuset if qmax > queues_count_from_cpuset else qmax
+                if qcurrent != new_rx_queues_count:
+                    self.__set_driver_rx_queues(iface, new_rx_queues_count)
+                perftune_print("Distributing IRQs handling Rx after limiting the rx hw queues {} :".format(new_rx_queues_count))
+                distribute_irqs(all_irqs[0:new_rx_queues_count], self.irqs_cpu_mask)
```
Comment (vladzcloudius) [inline scripts/perftune.py:834]: You are at the wrong clause. The condition at the line 806 is detecting NICs that have a limitation on the max RSS count which is less than the total number of IRQs present.

The trimming Rx queues down logic covers all NICs and hence should be outside and before this clause.

In particular we should trim Rx queues, make sure that `max_num_rx_queues` reflects that and only then continue to `distribute_irqs` calls.

File: scripts/perftune.py [hunk 27]
```diff
@@ -1258,10 +1313,12 @@ def names():
  --nic NIC       - default: eth0
  --cpu-mask MASK - default: all available cores mask
  --tune-clock    - default: false
+ --tune-rxqueues    - default: false
```
Comment (vladzcloudius) [inline scripts/perftune.py:1328]: We should also add the ability to override the amount of Rx queues: the default should be 1 - per IRQ CPU as in this patch and a user should be able to set the total number of RX_QUEUEs he/she wants that then will be configured and caped by the max HW supported value as per `ethtool -l`.

File: scripts/perftune.py [hunk 28]
```diff
@@ -1302,6 +1359,9 @@ def parse_options_file(prog_args):
     if 'tune_clock' in y and not prog_args.tune_clock:
         prog_args.tune_clock= y['tune_clock']
 
+    if 'tune_rxqueues' in y and not prog_args.tune_clock:
```
Comment (vladzcloudius) [inline scripts/perftune.py:1374]: Looks like a typo: `...and not prog_args.tune_clock:`

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py [hunk 29]
```diff
@@ -614,14 +618,18 @@ def __dev_is_hw_iface(self, iface):
     def __check_dev_is_bond_iface(self):
         if not os.path.exists('/sys/class/net/bonding_masters'):
             return False
-
-        return any([re.search(self.nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        bonddict={}
```
Comment (vladzcloudius) [inline scripts/perftune.py:621]: spaces around "=". 
Something is terribly wrong with your editor style config ;).

File: scripts/perftune.py [hunk 30]
```diff
@@ -614,14 +618,18 @@ def __dev_is_hw_iface(self, iface):
     def __check_dev_is_bond_iface(self):
         if not os.path.exists('/sys/class/net/bonding_masters'):
             return False
```
Comment (vladzcloudius) [inline scripts/perftune.py:620]: This will be an issue.
You should be consistent with the type the method returns.
With your patch it is `dict{<NIC name> -> bool}`.

File: scripts/perftune.py [hunk 31]
```diff
@@ -614,14 +618,18 @@ def __dev_is_hw_iface(self, iface):
...
+            bonddict[nic]=any([re.search(nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        return bonddict
 
     def __learn_slaves(self):
-        if self.nic_is_bond_iface:
-            return list(itertools.chain.from_iterable([ line.split() for line in open("/sys/class/net/{}/bonding/slaves".format(self.nic), 'r').readlines() ]))
+        slaveslistpernic={}
+        for nic in self.nics:
+            if self.nic_is_bond_iface(nic):
+                slaveslistpernic[nic]=list(itertools.chain.from_iterable([ line.split() for line in open("/sys/class/net/{}/bonding/slaves".format(nic), 'r').readlines() ]))
```
Comment (vladzcloudius) [inline scripts/perftune.py:630]: Here it's the same style disaster again: missing spaces around "=" and extra spaces after "[" and before "]".

File: scripts/perftune.py [hunk 32]
```diff
@@ -614,14 +618,18 @@ def __dev_is_hw_iface(self, iface):
...
-        return any([re.search(self.nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        bonddict={}
+        for nic in self.nics:
+            bonddict[nic]=any([re.search(nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        return bonddict
 
     def __learn_slaves(self):
-        if self.nic_is_bond_iface:
-            return list(itertools.chain.from_iterable([ line.split() for line in open("/sys/class/net/{}/bonding/slaves".format(self.nic), 'r').readlines() ]))
+        slaveslistpernic={}
```
Comment (vladzcloudius) [inline scripts/perftune.py:627]: Please, use the same variable name convention as the rest of the script: 
 * Use "_" as delimiters of a variable name.
 * Use only lower case characters.

For instance `slaveslistpernic` should be `slaves_list_per_nic`

File: scripts/perftune.py [hunk 33]
```diff
@@ -614,14 +618,18 @@ def __dev_is_hw_iface(self, iface):
     def __check_dev_is_bond_iface(self):
         if not os.path.exists('/sys/class/net/bonding_masters'):
             return False
-
-        return any([re.search(self.nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        bonddict={}
+        for nic in self.nics:
+            bonddict[nic]=any([re.search(nic, line) for line in open('/sys/class/net/bonding_masters', 'r').readlines()])
+        return bonddict
```
Comment (vladzcloudius) [inline scripts/perftune.py:624]: s/bonddict/bond_dict/

File: scripts/perftune.py [hunk 34]
```diff
@@ -779,10 +787,14 @@ def __learn_irqs(self):
         This is a slow method that is going to read from the system files. Never
         use it outside the initialization code.
         """
-        if self.nic_is_bond_iface:
-            return { slave : self.__learn_irqs_one(slave) for slave in filter(self.__dev_is_hw_iface, self.slaves) }
-        else:
-            return { self.nic : self.__learn_irqs_one(self.nic) }
+        nicIrqDic={}
```
Comment (vladzcloudius) [inline scripts/perftune.py:790]: Also be consistent with suffixes: you already used `_dict` in similar circumstances above. 

s/nicIrqDic/nic_irq_dict/

File: scripts/perftune.py [hunk 35]
```diff
@@ -1387,7 +1399,7 @@ def dump_config(prog_args):
         prog_options['mode'] = prog_args.mode
 
     if prog_args.nic:
-        prog_options['nic'] = prog_args.nic
+        prog_options['nic'] = prog_args.nics
```
Comment (vladzcloudius) [inline scripts/perftune.py:1402]: And what about `parse_options_file()`? It should adjusted too.
The default case handling needs to be changed too: `args.nics` should be set to [ 'eth0' ] and the `args.nic` initialization needs to be killed.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: scripts/perftune.py (hunk 21 above)
Comment (tarzanek) [reply scripts/perftune.py:588]: addressed in the same commit
squashed with
"don't make changing ethtool as read only"

File: scripts/perftune.py (hunk 23 above)
Comment (tarzanek) [reply scripts/perftune.py:677]: fixed in the same commit, added comment:
comparator for mlx irqs should detect if mlx is not present

File: scripts/perftune.py (hunk 24 above)
Comment (tarzanek) [reply scripts/perftune.py:776]: fixed in original commit, added " set proper message for driver detection "

File: scripts/perftune.py [hunk 36]
```diff
@@ -689,6 +689,49 @@ def __get_driver_name(self, iface):
 
         return driver_name
 
+    def __set_driver_rx_queues(self, iface, queue_count):
+        """
+        :param iface: interface to set
+        :param queue_count: number of RX queues to set
+        """
+        try:
+                perftune_print("Limiting RX queues on {} ...".format(iface))
```
[the remark this answers is not in the document]
Comment (tarzanek) [reply scripts/perftune.py:708]: fixed in "set rx queues cleanup"

File: scripts/perftune.py [hunk 37]
```diff
@@ -689,6 +689,49 @@ def __get_driver_name(self, iface):
...
+                ethmodes = run_ethtool(['-l', iface])
```
[the remark this answers is not in the document]
Comment (tarzanek) [reply scripts/perftune.py:709]: fixed in "set rx queues cleanup"

File: scripts/perftune.py [hunk 38]
```diff
@@ -689,6 +689,49 @@ def __get_driver_name(self, iface):
...
+                combined=False
+                for line in ethmodes:
+                    splitline = line.split(":")
+                    id = splitline[0]
+                    count = splitline[1].strip()
+                    if id == "Combined" and int(count) > 0:
```
[the remark this answers is not in the document]
Comment (tarzanek) [reply scripts/perftune.py:715]: fixed in "set rx queues cleanup"

File: scripts/perftune.py (hunk 29 above)
Comment (tarzanek) [reply scripts/perftune.py:621]: I will run autoformat now always before commiting code (and will recheck all diffs with it), I will use default style from Visual Studio Code (or any other you point me to or if there will be any set in seastar)

**Comment by tarzanek** [comment]
I removed the code for setting rx queues and it will be part of a new PR

**Comment by tarzanek** [comment]
@avikivity can a seastar maintainer review and merge this?

File: scripts/perftune.py [hunk 39]
```diff
@@ -1366,7 +1384,10 @@ def dump_config(prog_args):
 
 # set default values #####################
 if not args.nic:
```
Comment (vladzcloudius) [inline scripts/perftune.py:1386]: I don't think args.nic can ever be not None after you changed the `dest` to be `args.nics`.
I think you should drop this whole if-then. You already have the corresponding handling of args.nics right below.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]
A small fix (cleanup) is required. 
Except for that looks good.

**Comment by tarzanek** [comment]
@avikivity will you have some time to review? Vlad already did a terrific job on review and I hope all the code is clear now

**Comment by tarzanek** [comment]
@penberg can you help with a review here? shouldn't be a hard one, patches should be formatted properly and Vlad already reviewed, tia :-)

File: scripts/perftune.py [hunk 40]
```diff
@@ -743,7 +749,7 @@ def __max_rx_queue_count(self, iface):
         driver_to_max_rss = {'ixgbe': 16, 'ixgbevf': 4, 'i40e': 64, 'i40evf': 16}
 
         driver_name = ''
-        ethtool_i_lines = run_read_only_command(['ethtool', '-i', iface]).splitlines()
+        ethtool_i_lines = run_ethtool(['-i', iface]).splitlines()
```
Comment (penberg) [inline scripts/perftune.py]: Isn't this `splitline()` redundant as `run_ethtool` already does it?

Is the commit message out of sync with the commit itself? This looks like a pure cleanup with no functional changes, no?

**Review by penberg — COMMENTED** [review]
Please add some text to the pull request to explain why this is needed (each NIC has their specific IRQ/queue mapping?). Also explain why we need support for passing multiple NICs.

There's also one minor comment about a commit message, which looks out of sync with the commit itself.

File: scripts/perftune.py (hunk 40 above)
Comment (tarzanek) [reply scripts/perftune.py]: squashed refactor of ethtool and adding __get_driver_name to fix this

**Comment by penberg** [comment]
Well, there was a merge commit, but while touching up the git commit messages, I managed to remove it...

**Comment by penberg** [comment]
Merged as of https://github.com/scylladb/seastar/commit/a2fc9d72c31b97840bc75ae49dbd6f4b6d394e25

**Comment by tarzanek** [comment]
ah ... maybe I should have rebased, I will rebase next time
